### PR TITLE
Trigger memory revoking quickly on over-allocation

### PR DESCRIPTION
**Pull request by findepi**
When MemoryPool is over-allocated, the MemoryRevokingScheduler's check
is asynchronously initiated for that pool.

### CODE REVIEW

**Comment by findepi** [comment]
@losipiuk, this is a WIP. Tests not updated yet (don't even compile now). Your feedback is welcome.

File: presto-main/src/main/java/com/facebook/presto/execution/MemoryRevokingScheduler.java [hunk 1]
```diff
@@ -15,39 +15,63 @@
...
+import java.util.function.Supplier;
 
+import static com.google.common.base.Preconditions.checkState;
+import static java.util.Collections.synchronizedSet;
 import static java.util.Objects.requireNonNull;
 
 public class MemoryRevokingScheduler
 {
-    private static final Logger LOG = Logger.get(MemoryRevokingScheduler.class);
+    private static final Logger log = Logger.get(MemoryRevokingScheduler.class);
```
Comment (losipiuk) [inline presto-main/src/main/java/com/facebook/presto/execution/MemoryRevokingScheduler.java]: Move rename to separate commit.

File: presto-main/src/main/java/com/facebook/presto/execution/MemoryRevokingScheduler.java [hunk 2]
```diff
@@ -56,17 +80,118 @@ public MemoryRevokingScheduler(LocalMemoryManager localMemoryManager, FeaturesCo
...
+        memoryPools.forEach(memoryPool -> memoryPool.addListener(memoryPoolListener));
+    }
+
+    @PreDestroy
+    public void stop()
+    {
+        memoryPools.forEach(memoryPool -> memoryPool.removeListener(memoryPoolListener));
+    }
+
+    void register(Supplier<? extends Collection<SqlTask>> currentTasks, ScheduledExecutorService taskManagementExecutor)
```
Comment (losipiuk) [inline presto-main/src/main/java/com/facebook/presto/execution/MemoryRevokingScheduler.java]: i would use currentTasksSupplier as variable name. It is harder to read it to me without the suffix.

File: presto-main/src/main/java/com/facebook/presto/execution/MemoryRevokingScheduler.java [hunk 3]
```diff
@@ -56,17 +80,118 @@ public MemoryRevokingScheduler(LocalMemoryManager localMemoryManager, FeaturesCo
...
+    private void checkMemoryPools(boolean checkAllPools)
+    {
+        /*
+         * There can be at most one thread actually executing this method. If pool or pools need to be re-checked,
+         * this will be signalled by setting immediateCheckPending and populating potentiallyOverflowingPools.
+         */
+        if (!checkRunning.compareAndSet(false, true)) {
+            // Already running
+
+            if (checkAllPools) {
```
Comment (losipiuk) [inline presto-main/src/main/java/com/facebook/presto/execution/MemoryRevokingScheduler.java]: why not do it outside the check-running block and then skip the `|| checkAllPools` part of condition below?

File: presto-main/src/main/java/com/facebook/presto/execution/MemoryRevokingScheduler.java [hunk 4]
```diff
@@ -56,17 +80,118 @@ public MemoryRevokingScheduler(LocalMemoryManager localMemoryManager, FeaturesCo
...
+            log.warn(e, "Error when acting on memory pool reservation");
+        }
     }
 
-    public void requestMemoryRevokingIfNeeded(Collection<SqlTask> sqlTasks)
+    @VisibleForTesting
+    void checkMemoryPools()
     {
-        memoryPools.forEach(memoryPool -> requestMemoryRevokingIfNeeded(sqlTasks, memoryPool));
+        checkMemoryPools(true);
```
Comment (losipiuk) [inline presto-main/src/main/java/com/facebook/presto/execution/MemoryRevokingScheduler.java]: Actually I do not like this rename. IMO the previous name was more informative.

File: presto-main/src/main/java/com/facebook/presto/execution/MemoryRevokingScheduler.java (hunk 3 above)
Comment (findepi) [reply presto-main/src/main/java/com/facebook/presto/execution/MemoryRevokingScheduler.java]: The reason is to avoid adding/removing the pools to the Set every 1 s when nothing else happens. I.e. to keep low profile when idle.

File: presto-main/src/main/java/com/facebook/presto/execution/MemoryRevokingScheduler.java [hunk 5]
```diff
@@ -51,9 +51,9 @@ public MemoryRevokingScheduler(LocalMemoryManager localMemoryManager, FeaturesCo
     }
```
Comment (pnowojski) [inline presto-main/src/main/java/com/facebook/presto/execution/MemoryRevokingScheduler.java:111]: Typo in commit message `but` -> `bug`

File: presto-main/src/test/java/com/facebook/presto/execution/TestMemoryRevokingScheduler.java [hunk 6]
```diff
@@ -58,17 +60,21 @@
 
 public class TestMemoryRevokingScheduler
 {
-    private final ScheduledExecutorService executor;
```
Comment (losipiuk) [inline presto-main/src/test/java/com/facebook/presto/execution/TestMemoryRevokingScheduler.java]: This possibly needs 
```java
@Test(singleThreaded = true)
```
annotation.

File: presto-main/src/main/java/com/facebook/presto/execution/MemoryRevokingScheduler.java [hunk 7]
```diff
@@ -51,9 +51,9 @@ public MemoryRevokingScheduler(LocalMemoryManager localMemoryManager, FeaturesCo
...
 
     @VisibleForTesting
-    MemoryRevokingScheduler(MemoryPool memoryPool, double memoryRevokingThreshold, double memoryRevokingTarget)
```
Comment (losipiuk) [inline presto-main/src/main/java/com/facebook/presto/execution/MemoryRevokingScheduler.java]: `s/but/bug/` in commit message

File: presto-main/src/test/java/com/facebook/presto/execution/TestMemoryRevokingScheduler.java [hunk 8]
```diff
@@ -170,6 +173,54 @@ public void testScheduleMemoryRevoking()
...
+         * sqlTask1 fills its pool
+         */
+        operatorContext1.setRevocableMemoryReservation(12);
+        scheduler.requestMemoryRevokingIfNeeded(tasks);
+        assertMemoryRevokingRequestedFor(operatorContext1);
+
+        /*
+         * When sqlTask2 fills its pool
+         */
+
```
Comment (losipiuk) [inline presto-main/src/test/java/com/facebook/presto/execution/TestMemoryRevokingScheduler.java]: remove empty line

File: presto-main/src/test/java/com/facebook/presto/execution/TestMemoryRevokingScheduler.java [hunk 9]
```diff
@@ -170,6 +173,54 @@ public void testScheduleMemoryRevoking()
         assertMemoryRevokingRequestedFor(operatorContext3, operatorContext4);
     }
 
+    @Test
+    public void testCountAlreadyRevokedMemoryWithinAPool()
+            throws Exception
+    {
+        // Given
+
```
Comment (losipiuk) [inline presto-main/src/test/java/com/facebook/presto/execution/TestMemoryRevokingScheduler.java]: remove empty line

File: presto-main/src/main/java/com/facebook/presto/execution/MemoryRevokingScheduler.java [hunk 10]
```diff
@@ -24,11 +24,15 @@
 import com.google.common.collect.Ordering;
 import io.airlift.log.Logger;
 
+import javax.annotation.PostConstruct;
```
Comment (losipiuk) [inline presto-main/src/main/java/com/facebook/presto/execution/MemoryRevokingScheduler.java]: Type in commit message: 
s/TaskManagementexecutor/TaskManagementExecutor/

File: presto-main/src/main/java/com/facebook/presto/execution/MemoryRevokingScheduler.java [hunk 11]
```diff
@@ -47,6 +53,14 @@
     private final double memoryRevokingThreshold;
     private final double memoryRevokingTarget;
 
+    private final MemoryPoolListener memoryPoolListener = MemoryPoolListener.onMemoryReserved(this::onMemoryReserved);
```
Comment (losipiuk) [inline presto-main/src/main/java/com/facebook/presto/execution/MemoryRevokingScheduler.java]: inline this into `registerPoolListeners`

File: presto-main/src/main/java/com/facebook/presto/execution/MemoryRevokingScheduler.java [hunk 12]
```diff
@@ -75,30 +90,122 @@ public MemoryRevokingScheduler(
         this.taskManagementExecutor = taskManagementExecutor;
         this.memoryRevokingThreshold = memoryRevokingThreshold;
         this.memoryRevokingTarget = memoryRevokingTarget;
+        this.potentiallyOverflowingPools = synchronizedSet(new HashSet<>());
+    }
+
+    private void onMemoryReserved(MemoryPool memoryPool)
```
Comment (losipiuk) [inline presto-main/src/main/java/com/facebook/presto/execution/MemoryRevokingScheduler.java]: move this below `registerPoolListeners`

File: presto-main/src/main/java/com/facebook/presto/execution/MemoryRevokingScheduler.java [hunk 13]
```diff
@@ -75,30 +90,122 @@ public MemoryRevokingScheduler(
...
+    void requestMemoryRevokingIfNeeded()
+    {
+        requestMemoryRevokingIfNeeded(true);
+    }
+
+    private void requestMemoryRevokingIfNeeded(boolean checkAllPools)
+    {
+        /*
+         * There can be at most one thread actually executing this method. If pool or pools need to be re-checked,
+         * this will be signalled by setting immediateCheckPending and populating potentiallyOverflowingPools.
```
Comment (losipiuk) [inline presto-main/src/main/java/com/facebook/presto/execution/MemoryRevokingScheduler.java]: there is no variable `immediateCheckPending`

File: presto-main/src/main/java/com/facebook/presto/execution/MemoryRevokingScheduler.java [hunk 14]
```diff
@@ -75,30 +90,122 @@ public MemoryRevokingScheduler(
...
 
     @PostConstruct
     public void start()
     {
+        registerPeriodCheck();
+        registerPoolListeners();
+    }
+
+    @VisibleForTesting
+    void registerPeriodCheck()
```
Comment (losipiuk) [inline presto-main/src/main/java/com/facebook/presto/execution/MemoryRevokingScheduler.java]: rename to `registerPeriodicCheck`

File: presto-main/src/test/java/com/facebook/presto/execution/TestMemoryRevokingScheduler.java (hunk 6 above)
Comment (findepi) [reply presto-main/src/test/java/com/facebook/presto/execution/TestMemoryRevokingScheduler.java]: Fixed

**Review by losipiuk — CHANGES_REQUESTED** [review]
generally ok. Minor comments.

**Comment by losipiuk** [comment]
@kokosing please take a look at the last commit. Maybe you will have idea how to make it simpler.

File: presto-main/src/main/java/com/facebook/presto/execution/MemoryRevokingScheduler.java (hunk 11 above)
Comment (findepi) [reply presto-main/src/main/java/com/facebook/presto/execution/MemoryRevokingScheduler.java]: I can't inline, because there also is stop() where I deregister the listener

File: presto-main/src/main/java/com/facebook/presto/execution/MemoryRevokingScheduler.java (hunk 14 above)
Comment (findepi) [reply presto-main/src/main/java/com/facebook/presto/execution/MemoryRevokingScheduler.java]: obviously a typo

File: presto-main/src/main/java/com/facebook/presto/execution/MemoryRevokingScheduler.java [hunk 15]
```diff
@@ -75,30 +90,122 @@ public MemoryRevokingScheduler(
...
+                 *
+                 * There is a small chance of a race here. If the thread currently executing checkMemoryPools()
+                 * already did its last check of potentiallyOverflowingPools.isEmpty() but still did not reset
+                 * checkRunning flag, the check won't be done at all -- until periodic check is re-run or immediate
+                 * check is triggered (whichever happens earlier). Note however, that in such a situation the delay
+                 * until re-check will be equal to periodic check scheduling frequency, i.e. in no way worse than
+                 * normally.
+                 */
+                synchronized (potentiallyOverflowingPools) {
+                    memoryPools.forEach(potentiallyOverflowingPools::add);
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/execution/MemoryRevokingScheduler.java]:
```
potentiallyOverflowingPools::addAll(memoryPools);
```

File: presto-main/src/main/java/com/facebook/presto/execution/MemoryRevokingScheduler.java [hunk 16]
```diff
@@ -75,30 +90,122 @@ public MemoryRevokingScheduler(
...
+                }
+            }
+
+            return;
+        }
+        try {
+            do {
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/execution/MemoryRevokingScheduler.java]: use `while`

File: presto-main/src/main/java/com/facebook/presto/execution/MemoryRevokingScheduler.java [hunk 17]
```diff
@@ -47,6 +53,14 @@
     private final double memoryRevokingThreshold;
     private final double memoryRevokingTarget;
 
+    private final MemoryPoolListener memoryPoolListener = MemoryPoolListener.onMemoryReserved(this::onMemoryReserved);
+
+    /**
+     * Ids of {@link MemoryPool}s that need checking.
+     */
+    private final Set<MemoryPool> potentiallyOverflowingPools;
+    private final AtomicBoolean checkRunning = new AtomicBoolean(false);
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/execution/MemoryRevokingScheduler.java]: s/check/is

File: presto-main/src/main/java/com/facebook/presto/execution/MemoryRevokingScheduler.java [hunk 18]
```diff
@@ -75,30 +90,122 @@ public MemoryRevokingScheduler(
...
+            if (!memoryRevokingNeeded(memoryPool, memoryPool.getFreeBytes())) {
+                return;
+            }
+
+            if (potentiallyOverflowingPools.add(memoryPool)) {
+                log.debug("Scheduling check for %s", memoryPool);
+
+                taskManagementExecutor.execute(() -> {
+                    try {
+                        requestMemoryRevokingIfNeeded(false);
```
Comment (findepi) [inline presto-main/src/main/java/com/facebook/presto/execution/MemoryRevokingScheduler.java]: There's a race here. The pool has been over-allocated, but the OperatorContext that called the pool potentially did not update it's revokable memory counter yet. If the `requestMemoryRevokingIfNeeded` is run too early, it might still revoke some memory (when there are multiple tasks with revokable memory or when the requesting one already had some revokable memory). Or, `requestMemoryRevokingIfNeeded` might see no revokable memory and will be void.

File: presto-main/src/main/java/com/facebook/presto/execution/MemoryRevokingScheduler.java (hunk 10 above)
Comment (losipiuk) [reply presto-main/src/main/java/com/facebook/presto/execution/MemoryRevokingScheduler.java]: A few other typos now :)

Exract
SqlTaskTManager

File: presto-main/src/main/java/com/facebook/presto/memory/MemoryPool.java [hunk 19]
```diff
@@ -70,56 +74,93 @@ public synchronized MemoryPoolInfo getInfo()
...
-        return NOT_BLOCKED;
+
+        onMemoryReserved();
+        return result;
     }
 
-    public synchronized ListenableFuture<?> reserveRevocable(QueryId queryId, long bytes)
+    private void onMemoryReserved()
+    {
+        listeners.forEach(l -> l.onMemoryReserved(this));
```
Comment (losipiuk) [inline presto-main/src/main/java/com/facebook/presto/memory/MemoryPool.java]: use `listener` as lambda parameter.
We avoid single letter acronyms in codebase even is such context.

**Review by losipiuk — APPROVED** [review]

File: presto-main/src/main/java/com/facebook/presto/execution/MemoryRevokingScheduler.java (hunk 18 above)
Comment (findepi) [reply presto-main/src/main/java/com/facebook/presto/execution/MemoryRevokingScheduler.java]: Fixed

File: presto-main/src/main/java/com/facebook/presto/execution/MemoryRevokingScheduler.java [hunk 20]
```diff
@@ -38,26 +42,54 @@
...
+    private final Supplier<? extends Collection<SqlTask>> currentTasksSupplier;
+    private final ScheduledExecutorService taskManagementExecutor;
     private final double memoryRevokingThreshold;
     private final double memoryRevokingTarget;
 
     @Inject
-    public MemoryRevokingScheduler(LocalMemoryManager localMemoryManager, FeaturesConfig config)
+    public MemoryRevokingScheduler(
+            LocalMemoryManager localMemoryManager,
+            SqlTaskManager sqlTaskManager,
```
Comment (losipiuk) [inline presto-main/src/main/java/com/facebook/presto/execution/MemoryRevokingScheduler.java:69]: Injecting `SqlTaskManager` will not work as it is not bound explicitly.
Currently it is bound to interface `TaskManager` which does not have `getAllTasks` method.
You have to either:

 push the method up (which is probably ugly as exposes `SqlTask` which seems somewhat specific to `SqlTaskManager` via `TaskManager`.

or:

Explicitly bind `SqlTaskManager` and then bind `TaskManager` to that previous binding.

I would go for latter

**Review by losipiuk — CHANGES_REQUESTED** [review]
One more comment on binding

File: presto-main/src/test/java/com/facebook/presto/execution/TestMemoryRevokingScheduler.java [hunk 21]
```diff
@@ -103,7 +106,7 @@ public void testScheduleMemoryRevoking()
             throws Exception
     {
         // todo test for different threshold/target
-        MemoryRevokingScheduler scheduler = new MemoryRevokingScheduler(memoryPool, 1.0, 1.0);
+        MemoryRevokingScheduler scheduler = new MemoryRevokingScheduler(singletonList(memoryPool), 1.0, 1.0);
```
Comment (pnowojski) [inline presto-main/src/test/java/com/facebook/presto/execution/TestMemoryRevokingScheduler.java]: If it's not supposed to be modified, we are more often using `ImmutableList.of()`

File: presto-main/src/main/java/com/facebook/presto/operator/OperatorContext.java [hunk 22]
```diff
@@ -84,7 +86,8 @@
     private final AtomicLong finishUserNanos = new AtomicLong();
 
     private final AtomicLong memoryReservation = new AtomicLong();
-    private final AtomicLong revocableMemoryReservation = new AtomicLong();
+    @GuardedBy("this")
```
Comment (pnowojski) [inline presto-main/src/main/java/com/facebook/presto/operator/OperatorContext.java:89]: I don't like the idea, that part of the class is `ThreadSafe` and part is not. 

Maybe we should decompose it? Into `OperatorContext` and `RevocableMemoryContext`. The later one would be thread safe, and it would be used both by `OperatorContext` and `MemoryRevokingScheduler`?

File: presto-main/src/main/java/com/facebook/presto/memory/MemoryPoolListener.java [hunk 23]
```diff
@@ -0,0 +1,41 @@
...
+     * Invoked when memory reservation completes successfully.
+     *
+     * @param memoryPool the {@link MemoryPool} where the reservation took place
+     */
+    default void onMemoryReserved(MemoryPool memoryPool) {}
+
+    /**
+     * Creates {@link MemoryPoolListener} implementing {@link #onMemoryReserved(MemoryPool)} only.
+     */
+    static MemoryPoolListener onMemoryReserved(Consumer<? super MemoryPool> action)
```
Comment (pnowojski) [inline presto-main/src/main/java/com/facebook/presto/memory/MemoryPoolListener.java:30]: Why do you need this static builder? Couldn't you inline this?

If not, rename this method to `create` or `createMemoryPoolListener`, which is both a convention in presto and is less confusing (since there are so many methods named `onMemoryReserved` - this static method, interface method and it's overloads and private method somewhere else)

File: presto-main/src/main/java/com/facebook/presto/execution/MemoryRevokingScheduler.java (hunk 5 above)
Comment (findepi) [reply presto-main/src/main/java/com/facebook/presto/execution/MemoryRevokingScheduler.java:111]: already fixed?

File: presto-main/src/main/java/com/facebook/presto/memory/MemoryPoolListener.java (hunk 23 above)
Comment (findepi) [reply presto-main/src/main/java/com/facebook/presto/memory/MemoryPoolListener.java:30]: The name collision is not unintentional. Intent is:

* be able to instantiate interface with a lamba / method expr, almost as if it was `@Fluent`
* but still be able to evolve the interface by adding new methods

So the name of this method should include "`onMemoryReserved`" to convey the returned implementation will implement the `onMemoryReserved` method and that method only. Can you propose some name that satisfies this?

File: presto-main/src/main/java/com/facebook/presto/operator/OperatorContext.java (hunk 22 above)
Comment (findepi) [reply presto-main/src/main/java/com/facebook/presto/operator/OperatorContext.java:89]: disregarded for now

**Comment by findepi** [comment]
will be superseded by rebased PR
